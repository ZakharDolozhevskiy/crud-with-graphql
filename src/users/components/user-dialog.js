import React from 'react';
import moment from 'moment';
import styled from 'styled-components';

import Button from '@material-ui/core/Button';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContentText from '@material-ui/core/DialogContentText';
import FormHelperText from '@material-ui/core/FormHelperText';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';

export class UserDetailsDialog extends React.PureComponent {
  constructor(props) {
    super();
    this.state = { ...props.data };
  }

  onNameChange = event =>
    this.setState({ name: event.target.value });

  onEmailChange = event =>
    this.setState({ email: event.target.value });

  onActiveStateChange = (ev, checked) =>
    this.setState({ active: checked });

  onSave = () => this.props.onSave(this.state);

  render() {
    const { error, title, onCancel } = this.props;
    const { createdAt, updatedAt, email, name, active } = this.state;

    return (
      <Dialog open onClose={onCancel}>
        <DialogTitle>{title}</DialogTitle>
        <DialogContent>
          <div className={this.props.className}>
            <FormControl>
              <InputLabel htmlFor="edit-user-name">
                Name
              </InputLabel>
              <Input
                value={name}
                id="edit-user-name"
                onChange={this.onNameChange}
              />
            </FormControl>
            <FormControl error={error}>
              <InputLabel htmlFor="edit-user-email">
                E-mail
              </InputLabel>
              <Input
                value={email}
                id="edit-user-email"
                onChange={this.onEmailChange}
              />
              {error && <FormHelperText>Email address already in use</FormHelperText>}
            </FormControl>
            <FormControlLabel
              label="Active user"
              control={
                <Checkbox
                  color="primary"
                  checked={active}
                  onChange={this.onActiveStateChange}
                />}
            />
            <DialogContentText className="timestamps">
              {createdAt && <span>Created: {moment(createdAt).fromNow()}</span>}
              {updatedAt && <span>Updated: {moment(updatedAt).fromNow()}</span>}
            </DialogContentText>
          </div>
        </DialogContent>
        <DialogActions>
          <Button onClick={onCancel} color="primary">Cancel</Button>
          <Button onClick={this.onSave} color="primary">Save</Button>
        </DialogActions>
      </Dialog>
    );
  }
}

export default styled(UserDetailsDialog)`
  width: 400px;
  
  & > div {
    display: flex;
    padding-bottom: 16px;
  }
  
  .timestamps > span {
    display: block;
  }
`;