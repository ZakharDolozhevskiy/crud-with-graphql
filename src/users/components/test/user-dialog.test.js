import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

import { user } from '../../test/__mocks__';
import { UserDetailsDialog } from '../user-dialog';

describe('<UserDetailsDialog/>', () => {
  const onSave = jest.fn();
  const onCancel = jest.fn();

  const component = shallow(
    <UserDetailsDialog
      title="x"
      data={user}
      onSave={onSave}
      onCancel={onCancel}
    />
  );

  it('renders dialog with user data', () => {
    expect(component.state()).toEqual(user);
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('renders dialog with error message', () => {
    component.setProps({ error: true });
    expect(toJSON(component)).toMatchSnapshot();
  });
});