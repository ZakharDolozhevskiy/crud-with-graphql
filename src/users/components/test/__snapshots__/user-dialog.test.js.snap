// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDetailsDialog/> renders dialog with error message 1`] = `
<WithStyles(Dialog)
  onClose={[Function]}
  open={true}
>
  <WithStyles(DialogTitle)>
    x
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <div>
      <WithStyles(FormControl)>
        <WithStyles(InputLabel)
          htmlFor="edit-user-name"
        >
          Name
        </WithStyles(InputLabel)>
        <WithStyles(Input)
          id="edit-user-name"
          onChange={[Function]}
          value="User"
        />
      </WithStyles(FormControl)>
      <WithStyles(FormControl)
        error={true}
      >
        <WithStyles(InputLabel)
          htmlFor="edit-user-email"
        >
          E-mail
        </WithStyles(InputLabel)>
        <WithStyles(Input)
          id="edit-user-email"
          onChange={[Function]}
          value="test@email"
        />
        <WithStyles(FormHelperText)>
          Email address already in use
        </WithStyles(FormHelperText)>
      </WithStyles(FormControl)>
      <WithStyles(FormControlLabel)
        control={
          <WithStyles(Checkbox)
            checked={true}
            color="primary"
            onChange={[Function]}
          />
        }
        label="Active user"
      />
      <WithStyles(DialogContentText)
        className="timestamps"
      >
        <span>
          Created: 
          0
        </span>
        <span>
          Updated: 
          1
        </span>
      </WithStyles(DialogContentText)>
    </div>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`<UserDetailsDialog/> renders dialog with user data 1`] = `
<WithStyles(Dialog)
  onClose={[Function]}
  open={true}
>
  <WithStyles(DialogTitle)>
    x
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <div>
      <WithStyles(FormControl)>
        <WithStyles(InputLabel)
          htmlFor="edit-user-name"
        >
          Name
        </WithStyles(InputLabel)>
        <WithStyles(Input)
          id="edit-user-name"
          onChange={[Function]}
          value="User"
        />
      </WithStyles(FormControl)>
      <WithStyles(FormControl)>
        <WithStyles(InputLabel)
          htmlFor="edit-user-email"
        >
          E-mail
        </WithStyles(InputLabel)>
        <WithStyles(Input)
          id="edit-user-email"
          onChange={[Function]}
          value="test@email"
        />
      </WithStyles(FormControl)>
      <WithStyles(FormControlLabel)
        control={
          <WithStyles(Checkbox)
            checked={true}
            color="primary"
            onChange={[Function]}
          />
        }
        label="Active user"
      />
      <WithStyles(DialogContentText)
        className="timestamps"
      >
        <span>
          Created: 
          0
        </span>
        <span>
          Updated: 
          1
        </span>
      </WithStyles(DialogContentText)>
    </div>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
